x-pack/libbeat/management/managerV2.go

```diff
@@ -0,0 +1,336 @@
+
+func (cm *BeatV2Manager) unitListen() {
+	cm.logger.Debugf("Listening for agent unit changes")
+	for {
+		select {
+		// The stopChan channel comes from the Manager interface Stop() method
+		case <-cm.stopChan:
+			cm.stopBeat()
+		case change := <-cm.client.UnitChanges():
+			switch change.Type {
+			// Within the context of how we send config to beats, I'm not sure there is a difference between
+			// A unit add and a unit change, since either way we can't do much more than call the reloader
+			case client.UnitChangedAdded:
+				// At this point we also get a log level, however I'm not sure the beats core logger provides a
+				// clean way to "just" change the log level, without resetting the whole log config
+				cm.logger.Debugf("Got unit added: %s", change.Unit.ID())
+				go cm.handleUnitReload(change.Unit)
+
+			case client.UnitChangedModified:
+				state, _, _ := change.Unit.Expected()
+				cm.logger.Debugf("Got unit modified: %s, expected state is %s", change.Unit.ID(), state)
+				// I'm assuming that a state STOPPED just tells us to shut down the entire beat,
+				// as such we don't really care about updating via a particular unit
+				if state == client.UnitStateStopped {
+					cm.stopBeat()
+				} else {
+					go cm.handleUnitReload(change.Unit)
+				}
+
+			case client.UnitChangedRemoved:
+				cm.logger.Debugf("Got unit removed: %s", change.Unit.ID())
+				cm.deleteUnit(change.Unit)
+			}
+		}
+
+	}
+}
+
+func (cm *BeatV2Manager) stopBeat() {
+	if !cm.isRunning {
+		return
+	}
+
+	// will we ever get a Unit removed for anything other than the main beat?
+	// Individual reloaders don't have a "stop" function, so the most we can do
+	// is just shut down a beat, I think.
+	cm.logger.Debugf("Stopping beat")
+	// stop the "main" beat runtime
+	unit, mainExists := cm.getMainUnit()
+	if mainExists {
+		_ = unit.UpdateState(client.UnitStateStopping, "stopping beat", nil)
+	}
+
+	cm.isRunning = false
+	cm.stopMut.Lock()
+	defer cm.stopMut.Unlock()
+	if cm.stopFunc != nil {
+		// I'm not 100% sure the once here is needed,
+		// but various beats tend to handle this in a not-quite-safe way
+		cm.beatStop.Do(cm.stopFunc)
+	}
+	cm.client.Stop()
+
+	if mainExists {
+		_ = unit.UpdateState(client.UnitStateStopped, "stopped beat", nil)
+	}
+
+}
+
+func (cm *BeatV2Manager) handleUnitReload(unit *client.Unit) {
+	cm.addUnit(unit)
+	unitType := unit.Type()
+
+	if unitType == client.UnitTypeOutput {
+		cm.handleOutputReload(unit)
+	} else if unitType == client.UnitTypeInput {
+		cm.handleInputReload(unit)
+	}
+}
+
+// Handle the updated config for an output unit
+func (cm *BeatV2Manager) handleOutputReload(unit *client.Unit) {
+	_, _, rawConfig := unit.Expected()
+	cm.logger.Debugf("Got Output unit config: %s", rawConfig.Type)
+
+	reloadConfig, err := groupByOutputs(rawConfig)
+	if err != nil {
+		errString := fmt.Errorf("Failed to generate config for output: %w", err)
+		_ = unit.UpdateState(client.UnitStateFailed, errString.Error(), nil)
```

Comment:
I assume if any of these fails the agent will just send the configuration again as a natural retry mechanism? Can we document this with a comment?

Reply:
I honestly have no idea how agent responds to the failed states. @blakerouse ?

Reply:
The checkin being periodic should resolve this. There's not really anything we could do here besides retrying.